state/machine_ports_ops.go

```diff
@@ -207,6 +212,32 @@ func (op *openClosePortRangesOperation) lookupUnitEndpoints() error {
 	return nil
 }
 
+// validatePendingChanges ensures that the none of the pending open/close
+// entries specifies an endpoint that is not defined by the unit's charm
+// metadata.
+func (op *openClosePortRangesOperation) validatePendingChanges() error {
+	for unitName, pendingRangesByEndpoint := range op.mpr.pendingOpenRanges {
+		// Already verified; ignore error
+		appName, _ := names.UnitApplication(unitName)
+		for pendingEndpointName := range pendingRangesByEndpoint {
+			if pendingEndpointName != "" && !op.endpointsNamesByApp[appName].Contains(pendingEndpointName) {
```

Comment:
In what case is it valid for pendingEndpointName to be an empty string?

Reply:
We follow the same pattern as with the default binding for applications. 

An empty endpoint signifies that a list of port ranges applies to **all** (present _and_ future) application endpoints for this unit. When you get down to the firewaller level, `all endpoints == all machine subnets` so we effectively get the same ingress rules as we do with 2.8 and earlier jujus.

Reply:
Ack.  Is this well documented in the comments already?  If not, a comment here would be great.

Reply:
Yeap! In multiple locations (state and the facades). I have also changed pretty much all tests that manipulate ports (in the previous refactoring PRs) to pass a `allEndpoints` constant instead of an empty string to make it clear what it's all about.